### ISSUE TITLE
POSIX::strftime hangs on %Z under cygwin

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9267]
Migrated from [rt.perl.org#52126](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=52126) (status was 'resolved')

Searchable as RT52126$

**Comment by p5pRT** [comment]
### From eroode@barrack.com

#### Created by eroode@barrack.com



To reproduce this problem\, type the following command&#8203;:

&nbsp;    perl \-MPOSIX \-le 'print POSIX&#8203;::strftime\("%Z"\,0\,0\,0\,26\,2\,108\)'

I tested this on three cygwin systems\, all running perl 5\.8\.8\, and
on each\, the above command hung with the CPU pegged at 100%\.

I tested under Windows/Activestate perl 5\.8\.8\, and it worked fine\.
A friend tested under Mac OSX with perl 5\.8\.8\, and it worked fine\.
This friend also tested on cygwin with 5\.8\.8\, and it hung for him\.

My POSIX\.pm says its $VERSION is 1\.09\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.8:

Configured by rurban at Sun Jul  8 19:08:44 GMT 2007.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=cygwin, osvers=1.5.24(0.15642),
archname=cygwin-thread-multi-64int
    uname='cygwin_nt-5.1 reini 1.5.24(0.15642) 2007-01-31 10:57 i686
cygwin '
    config_args='-de -Dmksymlinks -Duse64bitint -Dusethreads
-Uusemymalloc -Doptimize=-O3 -Dman3ext=3pm -Dusesitecustomize
-Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_USE_SAFE_PUTENV -fno-strict-aliasing
-pipe -Wdeclaration-after-statement',
    optimize='-O3',
    cppflags='-DPERL_USE_SAFE_PUTENV -fno-strict-aliasing -pipe
-Wdeclaration-after-statement'
    ccversion='', gccversion='3.4.4 (cygming special, gdc 0.12, using
dmd 0.125)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8,
Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='ld2', ldflags =' -s -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-lgdbm -ldb -ldl -lcrypt -lgdbm_compat
    perllibs=-ldl -lcrypt -lgdbm_compat
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' -s'
    cccdlflags=' ', lddlflags=' -s -L/usr/local/lib'

Locally applied patches:
    CYG01 - hints.cygwin.sh ldflags -s
    CYG02 - lib-ExtUtils-Embed insensitive against leading \\s
    CYG03 - lib-Test-Harness-Straps $ENV{PERL5LIB} = ''
    CYG04 - major.version.cygwin.sh cygperl-5_8.dll and not
cygperl-5_8_x.dll
    CYG05 - add Win32CORE to core
    CYG07 - File-Spec-Cygwin-TMPDIR.patch
    Bug#38628 - allow legacy Cwd->cwd()
    Bug#40103 - File-Spec-case_tolerant.patch from 5.9.5


@INC for perl v5.8.8:
    /cygdrive/h/lib
    /usr/lib/perl5/5.8/cygwin
    /usr/lib/perl5/5.8
    /usr/lib/perl5/site_perl/5.8/cygwin
    /usr/lib/perl5/site_perl/5.8
    /usr/lib/perl5/site_perl/5.8
    /usr/lib/perl5/vendor_perl/5.8/cygwin
    /usr/lib/perl5/vendor_perl/5.8
    /usr/lib/perl5/vendor_perl/5.8
    .


Environment for perl v5.8.8:
    CYGWIN_ROOT=\cygwin
    HOME=/cygdrive/h
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
 
PATH=/usr/sbin:/usr/ccs/bin:/usr/local/bin:/cygdrive/h/bin:/usr/bin:/bin
:/usr/X11R6/bin:.:/cygdrive/c/Program Files/ActiveState Perl Dev Kit
7.0/bin:/cygdrive/c/WINDOWS/system32:/cygdrive/c/WINDOWS://Summ2/Perl/bi
n://Summ2/Perl/site/bin:/cygdrive/c/WINDOWS/System32/Wbem:/cygdrive/c/Pr
ogram Files/Common Files/GTK/2.0/bin:/cygdrive/c/Program
Files/GNU/GnuPG/pub:/cygdrive/c/Program Files/CVSNT/
    PERL5LIB=/cygdrive/h/lib
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @jdhedden

Roode\, Eric wrote&#8203;:
>  To reproduce this problem\, type the following command&#8203;:
>
>     perl \-MPOSIX \-le 'print POSIX&#8203;::strftime\("%Z"\,0\,0\,0\,26\,2\,108\)'
>
>  I tested this on three cygwin systems\, all running perl 5\.8\.8\, and
>  on each\, the above command hung with the CPU pegged at 100%\.
>
>  I tested under Windows/Activestate perl 5\.8\.8\, and it worked fine\.
>  A friend tested under Mac OSX with perl 5\.8\.8\, and it worked fine\.
>  This friend also tested on cygwin with 5\.8\.8\, and it hung for him\.

I tried this with blead @&#8203; 33571\, and 5\.10\.x @&#8203; 33572 under
Cygwin\, and it didn't hang\.  Therefore\, this seems to have
been fixed\.

Then again\, it also didn't print anything out for the
timezone\.  Reading the 'strftime' section in 'perldoc
POSIX'\, tells why&#8203;:

&nbsp;    The "Z" specifier is notoriously unportable since the names
&nbsp;    of timezones are non\-standard\. Sticking to the numeric
&nbsp;    specifiers is the safest route\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @steve-m-hay

Roode\, Eric" \(via RT\) wrote&#8203;:
> \# New Ticket Created by  "Roode\, Eric"
> \# Please include the string&#8203;:  \[perl \#52126\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=52126 >
> 
> 
> This is a bug report for perl from eroode@&#8203;barrack\.com\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> To reproduce this problem\, type the following command&#8203;:
> 
>     perl \-MPOSIX \-le 'print POSIX&#8203;::strftime\("%Z"\,0\,0\,0\,26\,2\,108\)'
> 
> I tested this on three cygwin systems\, all running perl 5\.8\.8\, and
> on each\, the above command hung with the CPU pegged at 100%\.
> 
> I tested under Windows/Activestate perl 5\.8\.8\, and it worked fine\.
> A friend tested under Mac OSX with perl 5\.8\.8\, and it worked fine\.
> This friend also tested on cygwin with 5\.8\.8\, and it hung for him\.

This is probably the bug that was fixed in \#29350&#8203;:

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/29350

\(That change was after 5\.8\.8 was released\.\)

**Comment by p5pRT** [comment]
### From @smpeters

On Wed\, Mar 26\, 2008 at 8&#8203;:03 AM\, via RT Roode\, Eric
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  "Roode\, Eric"
>  \# Please include the string&#8203;:  \[perl \#52126\]
>  \# in the subject line of all future correspondence about this issue\.
>  \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=52126 >
>
>
>  This is a bug report for perl from eroode@&#8203;barrack\.com\,
>  generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
>
>
>  \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>  \[Please enter your report here\]
>
>  To reproduce this problem\, type the following command&#8203;:
>
>     perl \-MPOSIX \-le 'print POSIX&#8203;::strftime\("%Z"\,0\,0\,0\,26\,2\,108\)'
>
>  I tested this on three cygwin systems\, all running perl 5\.8\.8\, and
>  on each\, the above command hung with the CPU pegged at 100%\.
>
>  I tested under Windows/Activestate perl 5\.8\.8\, and it worked fine\.
>  A friend tested under Mac OSX with perl 5\.8\.8\, and it worked fine\.
>  This friend also tested on cygwin with 5\.8\.8\, and it hung for him\.
>
>  My POSIX\.pm says its $VERSION is 1\.09\.
>

Actually\, its the missing $is\_dst \(or default \-1\) that's killing it\.
The following prints finishes\.\.\.

perl \-MPOSIX=strftime \-le 'print strftime\("%Z"\,0\,0\,0\,26\,2\,108\,\-1\,\-1\,0\)'

but the following doesn't

 perl \-MPOSIX=strftime \-le 'print strftime\("%Z"\,0\,0\,0\,26\,2\,108\,\-1\,\-1\,1\)'

It likely has to do with Cygwin's way of dealing with %Z\.  From the manpage

&nbsp;       \`%Z'
&nbsp;            The time zone name\.  If tm\_isdst is negative\, no output is
&nbsp;            generated\.  Otherwise\, the time zone name is based on the TZ
&nbsp;            environment variable\, as if by calling tzset\(\)\. \[tm\_isdst\]

Hopefully change \#29350 is in the 5\.8\.9 release\.

Steve Peters
steve@&#8203;fisharerojo\.org

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Mar 26\, 2008 at 10&#8203;:54&#8203;:05AM \-0500\, Steve Peters wrote&#8203;:

> Hopefully change \#29350 is in the 5\.8\.9 release\.

grep says yes\!

$ grep \-C 3 29350 Changes
&nbsp;             named "isa"\. \(That happens when importing "isa" from UNIVERSAL\,
&nbsp;             for example\.\) Fixes bug \#24824\.
&nbsp;             
&nbsp;             \[ 29350\]
&nbsp;             Fix infinite loop in Perl\_my\_strftime\(\) for failing strftime\(\)
&nbsp;     Branch&#8203;: maint\-5\.8/perl
&nbsp;           \! t/op/pat\.t


Nicholas Clark

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
